onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+    int input_index = optimizer_utils::IndexOfNodeInput(*lead_cast, *cast_input);
+    graph.RemoveEdge(producer->Index(), lead_cast->Index(), output_index, input_index);
+    if (consumers.empty()) {
+      auto& outputs = producer->MutableOutputDefs();
+      std::replace(outputs.begin(), outputs.end(), cast_input, cast_output);
+      graph.UpdateProducerNode(cast_output->Name(), producer->Index());
+    }
+  }
+  // Update consumer nodes
+  if (consumers.size()>0) {
+    int cast_output_index = optimizer_utils::IndexOfNodeOutput(*trail_cast, *cast_output);
+    for (Node* consumer : consumers) {
+      if (consumer != nullptr &&
+          std::find(removed_nodes.begin(), removed_nodes.end(), consumer->Index()) == removed_nodes.end()) {
+        auto& consumer_inputs = consumer->MutableInputDefs();
+        int input_index = optimizer_utils::IndexOfNodeInput(*consumer, *cast_output);
+        graph.RemoveEdge(trail_cast->Index(), consumer->Index(), cast_output_index, input_index);
+        std::replace(consumer_inputs.begin(), consumer_inputs.end(), cast_output, cast_input);
+        if (producer) {
+          graph.AddEdge(producer->Index(), consumer->Index(), output_index, input_index);
+        }
+      }
+    }
+    graph.UpdateConsumerNodes(cast_input->Name(), consumers);
+  }
+  for (auto cast : casts) {
+    graph_utils::RemoveNodeOutputEdges(graph, *cast);
+    removed_nodes.push_back(cast->Index());
+  }
+  return Status::OK();
+}
+
+// RemoveBackToBackCasts
+// Remove FLOAT and FLOAT16 casts back-to-back, either FLOAT->FLOAT16 or FLOAT16->FLOAT
+// Condition: The parent cast should have only one output
+static bool RemoveBackToBackCasts(Graph& graph,
+                                  std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                  const logging::Logger& logger)
+{
+  ORT_UNUSED_PARAMETER(logger);
+  bool modified = false;
+  for (Node& node : graph.Nodes()) {
+    if (std::find(removed_nodes.begin(), removed_nodes.end(), node.Index()) == removed_nodes.end()) {
+      bool is_fp = IsCastTo(node, TensorProto::FLOAT);
+      bool is_fp16 = IsCastTo(node, TensorProto::FLOAT16);
+      if (!is_fp && ! is_fp16) {
+        continue;
+      }
+      if (node.MutableOutputDefs().size() == 1) {
+        NodeArg* cast_output = node.MutableOutputDefs()[0];
+        for (Node* child : graph.GetMutableConsumerNodes(cast_output->Name())) {
+          if (std::find(removed_nodes.begin(), removed_nodes.end(), child->Index()) == removed_nodes.end()) {
+            bool is_child_fp = IsCastTo(*child, TensorProto::FLOAT);
+            bool is_child_fp16 = IsCastTo(*child, TensorProto::FLOAT16);
+            if ((is_fp && is_child_fp16) || (is_fp16 && is_child_fp)) {
+              // The parent and child cancell out
+              VLOGS(logger, 1) << "RemoveBackToBackCasts: Removed Cast nodes  " << node.Name() << " and " << child->Name();
+              RemoveCastNodes(graph, {&node, child}, removed_nodes);
```

Comment:
You are making an assumption here the input of first cast, and output of second cast are always of same type. 
What about this case? 
(fp32) -> CastTo32 -> CastTo16 -> (fp16)

Reply:
As I have explained (fp32)->CastTo32 would be removed at the top of the loop, because it is unnecessary.